src/overlay021/pokedexsort.c

```diff
@@ -0,0 +1,993 @@
+}
+
+BOOL PokedexSort_IsValidStep(UnkStruct_ov21_021D3320 *param0, int step)
+{
+    if (PokedexSort_IsValidDisplayIndex(param0, param0->sortedPokedex.currentDisplayIndex + step)) {
+        return TRUE;
+    }
+
+    return FALSE;
+}
+
+int PokedexSort_CurrentDisplayIndex(const UnkStruct_ov21_021D3320 *param0)
+{
+    return param0->sortedPokedex.currentDisplayIndex;
+}
+
+int PokedexSort_NumDisplayed(const UnkStruct_ov21_021D3320 *param0)
+{
+    return param0->sortedPokedex.numDisplayed;
+}
+
+int PokedexSort_DisplayIndexToStatusIndex(const UnkStruct_ov21_021D3320 *param0, int displayIndex)
+{
+    return param0->sortedPokedex.displayPokedex[displayIndex];
+}
+
+void PokedexSort_UpdateCurrentValues(UnkStruct_ov21_021D3320 *param0)
+{
+    for (int displayIndex = 0; displayIndex < param0->sortedPokedex.numDisplayed; displayIndex++) {
+        if (PokedexSort_DisplayIndexToStatusIndex(param0, displayIndex) == PokedexSort_CurrentStatusIndex(param0)) {
+            PokedexSort_SetCurrentValues(param0, displayIndex);
+            break;
+        }
+    }
+}
+
+int PokedexSort_NumEncountered(const UnkStruct_ov21_021D3320 *param0)
+{
+    return param0->numEncountered;
+}
+
+int PokedexSort_NumCaught(const UnkStruct_ov21_021D3320 *param0)
+{
+    return param0->numCaught;
+}
+
+Strbuf *PokedexSort_TrainerName(const UnkStruct_ov21_021D3320 *param0)
+{
+    return param0->trainerName;
+}
+
+u32 PokedexSort_TrainerGender(const UnkStruct_ov21_021D3320 *param0)
+{
+    return param0->trainerGender;
+}
+
+BOOL ov21_021D392C(const UnkStruct_ov21_021D3320 *param0, int param1)
+{
+    int species = PokedexSort_CurrentSpecies(param0);
+
+    if (sub_02027514(param0->dexData) == 0) {
+        return FALSE;
+    }
+
+    return sub_020274D0(param0->dexData, species, param1);
+}
+
+BOOL PokedexSort_CanDetectForms(const UnkStruct_ov21_021D3320 *param0)
+{
+    return Pokedex_CanDetectForms(param0->dexData);
+}
+
+void ov21_021D3960(UnkStruct_ov21_021D3320 *param0)
+{
+    if (gCoreSys.touchHeld) {
+        param0->unk_1758 = 0;
+        return;
+    }
+
+    if (gCoreSys.pressedKeys) {
+        if (param0->unk_1758 == 0) {
+            param0->unk_1758 = 1;
+        } else {
+            param0->unk_1758 = 2;
+        }
+    }
+}
+
+u32 ov21_021D3998(const UnkStruct_ov21_021D3320 *param0)
+{
+    return param0->unk_1758;
+}
+
+static void FilterUnencountered(u16 *encounteredDex, int *caughtStatusLength, const PokedexData *dexData, const u16 *fullDex, int pokedexLength)
+{
+    *caughtStatusLength = 0;
+
+    for (int species = 0; species < pokedexLength; species++) {
+        if (Pokedex_HasSeenSpecies(dexData, fullDex[species])) {
+            encounteredDex[*caughtStatusLength] = fullDex[species];
+            (*caughtStatusLength)++;
+        }
+    }
+}
+
+static void IntersectingPokedex(u16 *resultingPokedex, int *numResulting, const u16 *pokedex1, int dexLen1, const u16 *pokedex2, int dexLen2, BOOL keepUncaught, const PokedexData *dexData)
```

Comment:
I'm a bit confused on what this function does. It returns the intersection of the caught Pokemon in two different Pokedexes?

Reply:
Yes. So for example say you have caught Chimchar and a Bidoof and are searching for fire type pokemon. [chimchar, bidoof] will be one input and [charmander, charmelion, charizard, vulpix, ect.] will be the other. The resulting pokedex will be [chimchar]

Reply:
Gotcha. So this would then be used as an additional layer of filtering. Understood; thanks for the explanation! I think it would be nicer to name this as `IntersectPokedexes` or `CalcPokedexIntersection`, just to enforce the verbiage.